internal/rbd/rbd_util.go

```diff
@@ -902,58 +933,99 @@ func (rv *rbdVolume) getImageInfo(ctx context.Context, monitors string, cr *util
 	return imgInfo, nil
 }
 
-// snapInfo strongly typed JSON spec for snap ls rbd output
-type snapInfo struct {
-	ID        int64  `json:"id"`
-	Name      string `json:"name"`
-	Size      int64  `json:"size"`
-	Timestamp string `json:"timestamp"`
+func getCloneDepth(ctx context.Context, monitors, poolName, imageName string, depth int, cr *util.Credentials) (int, error) {
+	for {
+		if imageName == "" {
+			return depth, nil
+		}
+		vol := rbdVolume{
+			Pool:         poolName,
+			RbdImageName: imageName,
+		}
+		image, err := vol.getImageInfo(ctx, monitors, cr)
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to check depth on image: %s"), imageName, err)
+			return depth, err
+		}
+		if image.Parent.Image != "" {
+			depth++
+		}
+		imageName = image.Parent.Image
+	}
 }
 
-/*
-getSnapInfo queries rbd about the snapshots of the given image and returns its metadata, and
-returns ErrImageNotFound if provided image is not found, and ErrSnapNotFound if provided snap
-is not found in the images snapshot list
-*/
-func (rs *rbdSnapshot) getSnapInfo(ctx context.Context, monitors string, cr *util.Credentials) (snapInfo, error) {
-	// rbd --format=json snap ls [image-spec]
-
+func flattenRbdImage(ctx context.Context, rbdVol *rbdVolume, cr *util.Credentials) error {
 	var (
-		snpInfo snapInfo
-		snaps   []snapInfo
+		depth int
+		err   error
 	)
 
-	stdout, stderr, err := util.ExecCommand(
-		"rbd",
-		"-m", monitors,
-		"--id", cr.ID,
-		"--keyfile="+cr.KeyFile,
-		"-c", util.CephConfigPath,
-		"--format="+"json",
-		"snap", "ls", rs.imageSpec())
+	depth, err = getCloneDepth(ctx, rbdVol.Monitors, rbdVol.Pool, rbdVol.RbdImageName, 0, cr)
 	if err != nil {
-		klog.Errorf(util.Log(ctx, "failed getting snap (%s) information from image (%s): (%s)"),
-			rs.RbdSnapName, rs.imageSpec(), err)
-		if strings.Contains(string(stderr), "rbd: error opening image "+rs.RbdImageName+
-			": (2) No such file or directory") {
-			return snpInfo, ErrImageNotFound{rs.imageSpec(), err}
+		return err
+	}
+	klog.Infof(util.Log(ctx, "clone depth is (%d), configured softlimit (%v) and hardlimit (%v) for %s"), depth, rbdSoftMaxCloneDepth, rbdHardMaxCloneDepth, rbdVol)
+
+	if (depth != 0 && flattenImage) || (depth >= int(rbdHardMaxCloneDepth) || depth >= int(rbdSoftMaxCloneDepth)) {
```

Comment:
conversion of `rbdHardMaxCloneDepth` seems clutter, can these configuration options not be an int?